### PR TITLE
Refactoring and template trace generator

### PR DESCRIPTION
**Pull request by stoewer**
### What this PR does:

The purpose of this change is create more realistiacally looking traces with `xk6-client-tracing`.
* Refactored the code and introduce the concept of different trace generators. The existing functionality to generate traces is now provided by `tracing.ParameterizedGenerator`.
* Implement a new generator that creates traces and spans from a configurable template.

### Which issue(s) this PR fixes:

May help to solve grafana/tempo#902 because `xk6-client-tracing` could be used as a replacement for the `synthetic-load-generator`

### CODE REVIEW

**Review by mapno — COMMENTED** [review]
Like the direction of the PR, nice work. Left a few comments

File: README.md [hunk 1]
```diff
@@ -6,6 +6,121 @@
...
+
+```javascript
+{
+    // The defaults can be used to configure parameters that are applied to all spans (optional)
+    defaults: {
+        // Fixed attributes that are added to every generated span (optional)
+        attributes: { string : any },
+        // attributeSemantics can be set in order to generate attributes that follow a certain OpenTelemetry 
+        // semantic convention. For example tracing.SEMANTICS_HTTP (optional)
+        attributeSemantics: int,
```
[the remark this answers is not in the document]
Comment (stoewer) [reply README.md]: I changed it to a string type. I'm not sure if an array makes sense as for the purpose of this generator it makes sense when its one or the other i.e. is this an http request or a request to a database or message queue.

File: README.md [hunk 2]
```diff
@@ -6,6 +6,121 @@
...
+The configuration is an object with the following schema:
+
+```javascript
+{
+    // The endpoint to which the traces are sent in the form of <hostname>:<port>
+    endpoint: string,
+    // The exporter protocol used for sending the traces: tracing.EXPORTER_OTLP or tracing.EXPORTER_JAEGER
+    exporter: string,
+    // Whether insecure connections are allowed (optional, default: false)
+    insecure: bool
```
[the remark this answers is not in the document]
Comment (stoewer) [reply README.md]: Good point, I added the missing documentation.

File: tracing.go [hunk 3]
```diff
@@ -21,73 +20,147 @@ import (
...
+			"SEMANTICS_HTTP":  tracegen.SemanticsHTTP,
+			"SEMANTICS_DB":    tracegen.SemanticsDB,
+			"EXPORTER_OTLP":   exporterOTLP,
+			"EXPORTER_JAEGER": exporterJaeger,
+			// constructors
+			"Client":                 ct.newClient,
+			"ParameterizedGenerator": ct.newParameterizedGenerator,
+			"TemplatedGenerator":     ct.newTemplatedGenerator,
+			// functions
 			"generateRandomTraceID": ct.generateRandomTraceID,
```
[the remark this answers is not in the document]
Comment (stoewer) [reply tracing.go]: Agreed, I removed it.

File: pkg/tracegen/parameterized.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	RandomServiceName bool       `json:"random_service_name"`
+	Spans             SpanParams `json:"spans"`
+}
+
+type SpanParams struct {
+	Count      int                    `json:"count"`
+	Size       int                    `json:"size"`
+	RandomName bool                   `json:"random_name"`
+	FixedAttrs map[string]interface{} `json:"fixed_attrs"`
+}
```
[the remark this answers is not in the document]
Comment (stoewer) [reply pkg/tracegen/parameterized.go:31]: I added defaults for count and size. When the `TraceParams.ID` is not provided an ID will be randomly generated.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mapno) [reply README.md]: Agree. I wrote the comment before seeing the possible values. It makes sense like this.

**Review by mapno — APPROVED** [review]
I'm good to merge. I'll wait to see if @dgzlopes has any thoughts.

**Review by dgzlopes — APPROVED** [review]
LGTM! Thanks for the work @stoewer :bow:

**Pull request merged**